### PR TITLE
Feat: Login & Register

### PR DESCRIPTION
**Pull request by abdul-moize**
This branch adds the templates of login-register & header component to main branch

To test this PR. follow these steps

1. pull branch and run: python manage.py migrate && python manage.py runserver
2. pull this branch & run: npm start
3. Open http://localhost:3000/ in browser

Browse the site and let me if there are any issues, comments & concerns

### CODE REVIEW

File: src/pages/loginRegister.css [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+}
+.toggle-btn{
+  padding: 10px 30px;
+  cursor: pointer;
+  background: transparent;
+  border: 0;
+  outline: none;
+  position: relative;
+}
+#btn{
```
Comment (srehaider) [inline src/pages/loginRegister.css:32]: Assign classes to elements to style them, using Ids increases specificity which means if you will try to override styles using a class it won't work.

File: src/pages/loginRegister.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { useState } from "react"
+import "./loginRegister.css"
+import { LoginForm } from "../components/loginForm"
+import { RegisterForm } from "../components/registerForm"
+
+function LoginRegister(props) {
```
Comment (srehaider) [inline src/pages/loginRegister.js:6]:
```suggestion
function LoginOrRegisterForm(props) {
```

File: src/pages/loginRegister.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  const loginForm = <LoginForm setLoggedIn={props.setLoggedIn} />
+  const [form, setForm] = useState(loginForm)
```
Comment (srehaider) [inline src/pages/loginRegister.js:8]: Saving JSX into the state is not recommended instead, save values in the state and take decisions based on values.

```
const [activeFormType, setActiveFormType] = useState('login')
```

File: src/pages/loginRegister.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  const [button, setButton] = useState('0px')
```
Comment (srehaider) [inline src/pages/loginRegister.js:9]: What is the purpose of the button state?

File: src/pages/logout.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  localStorage.removeItem("token")
+  localStorage.removeItem("role")
+  const history = useHistory()
+  useEffect(() => {
+    history.replace(auth)
+    props.setLoggedIn(false)
+  }, [history, props])
+
+  return (<div></div>)
+}
```
Comment (srehaider) [inline src/pages/logout.js:15]: Move all logic related to logOut into the user service, once logout is done then redirect a user and this all logic should reside in `useEffect` which only once triggers when the component has been mounted.

File: src/components/registerForm.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+} from "./elements";
+
+function RegisterForm(props) {
+  const [nameError, setNameError] = useState();
+  const [emailError, setEmailError] = useState();
+  const [passwordError, setPasswordError] = useState();
+  const formRef = useRef();
+  const register = () => {
+    let form = new FormData(formRef.current);
+    registerService(form)
```
Comment (srehaider) [inline src/components/registerForm.js:18]: Same leave a blank line before and after this event handler
```suggestion
    let formData = new FormData(formRef.current);
    registerService(formData)
```

File: src/components/registerForm.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        setPasswordError([]);
+        props.changePage("Account created. Please log in");
+      })
+      .catch((errors) => {
+        setNameError("");
+        setEmailError("");
+        setPasswordError([]);
+        if ("name" in errors) setNameError(errors["name"]);
+        if ("email" in errors) setEmailError(errors["email"]);
+        if ("password" in errors) setPasswordError(errors["password"]);
```
Comment (srehaider) [inline src/components/registerForm.js:31]: Setting state is an expensive operation it rerenders screen in newer versions it has been optimized still it's better to use it only when required.
```suggestion
        setNameError("name" in errors ? errors["name"] : "");
        setEmailError("email" in errors ? errors["email"] : "");
        setPasswordError("password" in errors ? errors["password"] : []);
```

File: src/components/utils.js [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export function isContentManager() {
+  const roles = ["CM", "SA"];
+  return roles.some((value) => value === localStorage.getItem("role"));
+}
```
Comment (srehaider) [inline src/components/utils.js:4]: `utils` shouldn't reside inside `components` create a separate directory for it.

`roles` seems to be a constant value move it outside this function. Reading from `localSrorage` is an expensive operation you can read the value from `localStorage` and save it in a variable to compare with other values.
```suggestion
const CONTENT_MANAGER_ROLES = ["CM", "SA"];

export function isContentManager() {
  const current_user_role = localStorage.getItem("role");
  return CONTENT_MANAGER_ROLES.some((role) => role === current_user_role);
}
```

File: src/pages/loginOrRegister.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import { RegisterForm } from "../components/registerForm";
+
+function LoginOrRegister(props) {
+  const [activeForm, setActiveForm] = useState("login");
+  const [message, setMessage] = useState("");
+
+  let showLoginForm = (m = "") => {
+    setActiveForm("login");
+    if (typeof m === "string") setMessage(m);
+  };
```
Comment (srehaider) [inline src/pages/loginOrRegister.js:13]: This function is doing two jobs ideally, one function should do only one job. After reading the name it's hard to perceive it receives a message argument too. You can directly pass `setMessage` as a prop to setMessage from child components.

File: src/pages/loginOrRegister.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+  let showRegisterForm = () => {
+    setActiveForm("register");
+    setMessage("");
+  };
+  const loginForm = <LoginForm setLoggedIn={props.setLoggedIn} />;
+  const registerForm = <RegisterForm changePage={showLoginForm} />;
```
Comment (srehaider) [inline src/pages/loginOrRegister.js:20]: You are initializing both `LoginForm` and `RegisterForm` components and at the end, one will be discarded. You can initialize one component after taking a decision.

File: src/pages/loginOrRegister.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+          />
+          <button className="toggle-btn" onClick={showLoginForm}>
+            Log In
+          </button>
+          <button className="toggle-btn" onClick={showRegisterForm}>
+            Register
+          </button>
+        </div>
+        {message && <div className="message">{message}</div>}
+        {activeForm === "login" ? loginForm : registerForm}
```
Comment (srehaider) [inline src/pages/loginOrRegister.js:39]:
```suggestion
        {
          activeForm === "login" ?
          <LoginForm setLoggedIn={props.setLoggedIn} /> :
          <RegisterForm changePage={showLoginForm} />
        }
```

File: src/pages/logout.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import { AUTH } from "../components/constants";
+import { logout } from "../services/userService";
+
+function LogOut(props) {
+  const history = useHistory();
+  useEffect(() => {
+    logout();
+    history.replace(AUTH);
+    props.setLoggedIn(false);
+  }, [history, props]);
```
Comment (srehaider) [inline src/pages/logout.js:12]: `[history, props]` in `useEffect` means callback will be called every time `history` or `props` change. We don't want to do this we just want this callback to be invoked only once when the component mounts, to do so we pass an empty dependency array to `useEffect`.
```suggestion
  }, []);
```

File: src/services/userService.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+function AuthenticationException(message) {
+  this.message = message;
+}
+
+export const loginService = (form) => {
+  return fetch(LOGIN_API, {
+    method: "POST",
+    body: form,
+  })
```
Comment (srehaider) [inline src/services/userService.js:11]:
```suggestion
export const loginService = (loginCredentials) => {
  return fetch(LOGIN_API, {
    method: "POST",
    body: loginCredentials,
  })
```

File: src/services/userService.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    .catch((error) => {
+      throw error.message;
+    });
+};
+
+export const registerService = (form) => {
+  return fetch(REGISTER_API, {
+    method: "POST",
+    body: form,
+  })
```
Comment (srehaider) [inline src/services/userService.js:30]:
```suggestion
export const registerService = (userData) => {
  return fetch(REGISTER_API, {
    method: "POST",
    body: userData,
  })
```

File: src/components/loginForm.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+} from "./elements";
+
+function LoginForm(props) {
+  const history = useHistory();
+  const formRef = useRef();
+  const [errorMessage, setErrorMessage] = useState("");
+  const onSubmitHandler = (event) => {
+    event.preventDefault();
+    let form = new FormData(formRef.current);
+    loginService(form)
```
[the remark this answers is not in the document]
Comment (abdul-moize) [reply src/components/loginForm.js:19]: Done. If i leave a blank line after this call back then prettier complains.

File: .env [hunk 16]
```diff
@@ -0,0 +1,5 @@
+REACT_APP_PRODUCT_API = http://127.0.0.1:8000/products/api/
```
Comment (srehaider) [inline .env:1]: No need to move all API endpoints to the `.env` file, endpoints don't change with different environments only the domain does. You can move all endpoints to constants which will prepend domain from `.env`. We also don't commit the `.env` file every environment has its own local `.env` so put this file in `.gitignore` and commit another file `.env.example` with all required environment variables for reference.

**Review by srehaider — APPROVED** [review]
Looks great :+1: 
After moving API endpoints to constants and removing `.env` feel free to merge.

**Pull request merged**